src/swapRouter.ts

```diff
@@ -175,17 +179,173 @@ export abstract class SwapRouter {
+      const amountIn: string = toHex(trade.maximumAmountIn(options.slippageTolerance, inputAmount).quotient)
+      const amountOut: string = toHex(trade.minimumAmountOut(options.slippageTolerance, outputAmount).quotient)
+
+      // flag for whether the trade is single hop or not
+      const singleHop = route.pools.length === 1
+
+      const recipient =
+        typeof options.recipient === 'undefined' ? MSG_SENDER : validateAndParseAddress(options.recipient)
+      const mixedRouteIsAllV3 = (route: MixedRoute<Currency, Currency>) => {
+        return route.pools.every((pool) => pool instanceof Pool)
+      }
+
+      if (singleHop) {
+        // For single hop, since it isn't really a mixedRoute, we'll just mimic behavior of V3 or V2
+        if (route.pools.every((pool) => pool instanceof Pool)) {
+          /// it's guaranteed that the MixedRouteTrade here is compatible with V3Trade
+          return this.encodeV3Swap(
+            trade as unknown as V3Trade<Currency, Currency, TradeType>,
+            options,
+            routerMustCustody,
+            performAggregatedSlippageCheck
+          )
+        } else {
+          return [
+            ...calldatas,
+            this.encodeV2Swap(
+              trade as unknown as V2Trade<Currency, Currency, TradeType>,
+              options,
+              routerMustCustody,
+              performAggregatedSlippageCheck
+            ),
+          ]
+        }
+      } else {
+        let acc = []
+        let j = 0
+        while (j < route.pools.length) {
+          // seek forward until finding a pool of different type
+          let section = []
+          if (route.pools[j] instanceof Pool) {
+            while (route.pools[j] instanceof Pool) {
+              section.push(route.pools[j])
+              j++
+              if (j === route.pools.length) {
+                // we've reached the end of the route
+                break
+              }
+            }
+            acc.push(section)
+          } else {
+            while (route.pools[j] instanceof Pair) {
+              section.push(route.pools[j])
+              j++
+              if (j === route.pools.length) {
+                // we've reached the end of the route
+                break
+              }
+            }
+            acc.push(section)
+          }
+        }
+
+        const firstInputToken = route.input.wrapped
+        let outputToken = firstInputToken
+
+        const getOutputOfSection = (section: (Pool | Pair)[], firstInputToken: Token) => {
+          const { inputToken } = section.reduce(
+            ({ inputToken }, pool: Pool | Pair): { inputToken: Token } => {
+              const outputToken: Token = pool.token0.equals(inputToken) ? pool.token1 : pool.token0
+              return {
+                inputToken: outputToken,
+              }
+            },
+            { inputToken: firstInputToken }
+          )
+          return inputToken
+        }
```

Comment:
Shortened form of the logic in `encodeMixedRouteToPath` to get the output of a route containing multiple pools